File: typescript/tests/endpoint-allowlist.test.ts
/**
 * Endpoint resolution — explicit-config-only (security hardening).
 *
 * The endpoint is sourced ONLY from explicit in-code config (`resolveEndpoint`
 * arg), never from the process environment. An attacker who controls the env
 * (misconfigured CI runner, hostile container) can no longer set
 * `DEXCOST_ENDPOINT=http://attacker/` to exfiltrate cost telemetry or the
 * Bearer API key, because the SDK never reads that var.
 *
 * The explicit value is developer-supplied/trusted: it is honoured if it starts
 * with http:// or https:// (http:// is allowed for local/e2e since it is not
 * env-controllable); otherwise we fall back to the default with a warning.
 */

import { afterEach, describe, expect, test, vi } from "vitest";

import { resolveEndpoint, DEFAULT_ENDPOINT } from "../src/core/tracker.js";

const ORIGINAL_ENV = process.env.DEXCOST_ENDPOINT;

describe("endpoint resolution (explicit config only)", () => {
  afterEach(() => {
    if (ORIGINAL_ENV === undefined) {
      delete process.env.DEXCOST_ENDPOINT;
    } else {
      process.env.DEXCOST_ENDPOINT = ORIGINAL_ENV;
    }
    vi.restoreAllMocks();
  });

  test("returns the default when no explicit value is given", () => {
    expect(resolveEndpoint()).toBe(DEFAULT_ENDPOINT);
    expect(resolveEndpoint(DEFAULT_ENDPOINT)).toBe("https://api.dexcost.io");
  });

  test("treats an empty explicit value as unset", () => {
    expect(resolveEndpoint("")).toBe(DEFAULT_ENDPOINT);
  });

  test("honours an explicit https:// value", () => {
    expect(resolveEndpoint("https://custom.example")).toBe("https://custom.example");
  });

  test("honours an explicit http:// value (trusted, not env-controllable)", () => {
    // http:// is intentionally accepted for the explicit option (e.g.
    // http://localhost for e2e) — safe because it cannot come from the env.
    expect(resolveEndpoint("http://localhost:3000")).toBe("http://localhost:3000");
  });

  test("falls back to the default for a non-http(s) value with a warning", () => {
    const warnSpy = vi.spyOn(console, "warn").mockImplementation(() => {});
    expect(resolveEndpoint("javascript:alert(1)")).toBe(DEFAULT_ENDPOINT);
    expect(warnSpy).toHaveBeenCalled();
    expect(warnSpy.mock.calls[0]?.[0]).toContain("endpoint");
  });

  test("IGNORES the DEXCOST_ENDPOINT env var entirely", () => {
    // Even a malicious env value must have zero effect on resolution.
    process.env.DEXCOST_ENDPOINT = "http://evil.example";
    expect(resolveEndpoint()).toBe(DEFAULT_ENDPOINT);
    // And it never gets honoured implicitly via the explicit path either.
    expect(resolveEndpoint(undefined)).toBe(DEFAULT_ENDPOINT);
  });
});

File: go/endpoint_allowlist_test.go
// Endpoint validation — security/sdk-endpoint-explicit-config.
//
// The endpoint comes ONLY from explicit in-code Config.Endpoint; the SDK no
// longer reads the DEXCOST_ENDPOINT env var. Config.Endpoint is
// developer-supplied/trusted, so validation is minimal: a non-empty value
// must carry an http:// or https:// scheme (http:// is intentionally allowed
// for local e2e because it is not env-controllable). Anything else falls
// back to the production default.

package dexcost

import "testing"

func TestResolvedEndpoint_ExplicitHttpsAccepted(t *testing.T) {
	c := &Config{Endpoint: "https://custom.example.com"}
	if got := c.resolvedEndpoint(); got != "https://custom.example.com" {
		t.Errorf("expected https://custom.example.com, got %q", got)
	}
}

func TestResolvedEndpoint_ExplicitHttpLocalhostAccepted(t *testing.T) {
	// http:// is allowed for the explicit (trusted) field — e.g. local e2e.
	c := &Config{Endpoint: "http://localhost:3001"}
	if got := c.resolvedEndpoint(); got != "http://localhost:3001" {
		t.Errorf("expected http://localhost:3001, got %q", got)
	}
}

func TestResolvedEndpoint_RejectsSchemelessFallsBackToDefault(t *testing.T) {
	c := &Config{Endpoint: "ftp://x"}
	if got := c.resolvedEndpoint(); got != defaultEndpoint {
		t.Errorf("expected %q (default), got %q", defaultEndpoint, got)
	}
}

func TestResolvedEndpoint_RejectsArbitraryScheme(t *testing.T) {
	c := &Config{Endpoint: "javascript:alert(1)"}
	if got := c.resolvedEndpoint(); got != defaultEndpoint {
		t.Errorf("expected %q (default), got %q", defaultEndpoint, got)
	}
}

func TestResolvedEndpoint_DefaultWhenEmpty(t *testing.T) {
	c := &Config{}
	if got := c.resolvedEndpoint(); got != defaultEndpoint {
		t.Errorf("expected %q (default), got %q", defaultEndpoint, got)
	}
}
